File: app/src/main/java/com/github/overpass/gather/model/usecase/forgot/ForgotPasswordUseCase.java
package com.github.overpass.gather.model.usecase.forgot;

import androidx.lifecycle.LiveData;

import com.github.overpass.gather.model.commons.LiveDataUtils;
import com.github.overpass.gather.model.data.validator.BaseValidator;
import com.github.overpass.gather.model.repo.password.PasswordResetRepo;
import com.github.overpass.gather.screen.auth.login.forgot.ForgotStatus;

public class ForgotPasswordUseCase {

    private final PasswordResetRepo passwordResetRepo;
    private final BaseValidator validator;

    public ForgotPasswordUseCase(PasswordResetRepo passwordResetRepo, BaseValidator validator) {
        this.passwordResetRepo = passwordResetRepo;
        this.validator = validator;
    }

    public LiveData<ForgotStatus> sendForgotPassword(String email) {
        if (validator.isEmailValid(email)) {
            return passwordResetRepo.sendForgotPassword(email);
        } else {
            return LiveDataUtils.just(new ForgotStatus.InvalidEmail());
        }
    }
}

File: app/src/main/java/com/github/overpass/gather/model/usecase/forgot/ForgotPasswordUseCase.kt
package com.github.overpass.gather.model.usecase.forgot

import com.github.overpass.gather.di.EMAIL_VALIDATOR
import com.github.overpass.gather.model.commons.exception.InvalidCredentialsException
import com.github.overpass.gather.model.data.validator.Validator
import com.github.overpass.gather.model.repo.password.PasswordResetRepo
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import javax.inject.Inject
import javax.inject.Named

class ForgotPasswordUseCase @Inject constructor(
        private val passwordResetRepo: PasswordResetRepo,
        @Named(EMAIL_VALIDATOR) private val validator: Validator<String>
) {

    fun sendForgotPassword(email: String): Task<Void> {
        return email.takeIf { validator.isValid(it) }
                ?.let { passwordResetRepo.sendForgotPassword(it) }
                ?: Tasks.forException(InvalidCredentialsException("Invalid Email"))
    }
}
